app/code/core/Mage/Usa/Model/Shipping/Carrier/Ups.php

```diff
@@ -1584,13 +1455,403 @@ protected function _sendShipmentAcceptRequest(SimpleXMLElement $shipmentConfirmR
+        $shipToData = [
+            'Name'   => $request->getRecipientContactPersonName(),
+            'AttentionName' => $request->getRecipientContactPersonName(),
+            'Phone' => ['Number' => $request->getRecipientContactPhoneNumber()],
+            'Address' => [
+                'AddressLine' => $request->getRecipientAddressStreet1() . ' ' . $request->getRecipientAddressStreet2(),
+                'City' => $request->getRecipientAddressCity(),
+                'CountryCode' => $request->getRecipientAddressCountryCode(),
+                'PostalCode' => $request->getRecipientAddressPostalCode(),
+            ],
+        ];
+        if ($request->getRecipientAddressStateOrProvinceCode()) {
+            $shipToData['Address']['StateProvinceCode'] = $request->getRecipientAddressRegionCode();
+        }
+        if ($this->getConfigData('dest_type') == 'RES') {
+            $shipToData['Address']['ResidentialAddress'] = '';
+        }
+
+        if ($request->getIsReturn()) {
+            $shipFrom = &$shipParams['ShipmentRequest']['Shipment']['ShipFrom'];
+            $shipFrom['Name'] = $request->getShipperContactPersonName();
+            $shipFrom['AttentionName'] = $request->getShipperContactPersonName();
+            $address = &$shipFrom['Address'];
+            $address['AddressLine'] = $request->getShipperAddressStreet1() . ' ' . $request->getShipperAddressStreet2();
+            $address['City'] = $request->getShipperAddressCity();
+            $address['CountryCode'] = $request->getShipperAddressCountryCode();
+            $address['PostalCode'] = $request->getShipperAddressPostalCode();
+            if ($request->getShipperAddressStateOrProvinceCode()) {
+                $address['StateProvinceCode'] = $request->getShipperAddressStateOrProvinceCode();
+            }
+
+            $shipToAddress = &$shipToData['Address'];
+            $shipToAddress['AddressLine'] =
+                $request->getShipperAddressStreet1() . ' ' . $request->getShipperAddressStreet2();
+            $shipToAddress['City'] = $request->getShipperAddressCity();
+            $shipToAddress['CountryCode'] = $request->getShipperAddressCountryCode();
+            $shipToAddress['PostalCode'] = $request->getShipperAddressPostalCode();
+            if ($request->getShipperAddressStateOrProvinceCode()) {
+                $shipToAddress['StateProvinceCode'] = $request->getShipperAddressStateOrProvinceCode();
+            }
+            if ($this->getConfigData('dest_type') == 'RES') {
+                $shipToAddress['ResidentialAddress'] = '';
+            }
+        }
+
+        $shipParams['ShipmentRequest']['Shipment']['Service']['Code'] = $request->getShippingMethod();
+
+        $packagePart = [];
+        $customsTotal = 0;
+        $packagingTypes = [];
+        $deliveryConfirmationLevel = $this->_getDeliveryConfirmationLevel(
+            $request->getRecipientAddressCountryCode()
+        );
+        foreach ($packages as $packageId => $package) {
```

Comment:
Is it correct to loop over the packages here?

The function `_doShipmentRequest` is only called by the abstract base class (`Mage_Usa_Model_Shipping_Carrier_Abstract`, line 317). When it calls `_doShipmentRequest`, it's inside a loop over `$request->getPackages()` already, so this implementation will loop over all packages quadratically. The original XML implementation (`_doShipmentRequestXML`=>`_formShipmentRequest` just looks for the "current" package on the $request object that is set by the base class, e.g., `$request->getPackageId()`, `$request->getPackageParams()`, etc.

Reply:
The M2 UPS Carrier class also overrides the parent class's `requestToShipment` and `returnOfShipment` methods. The for-loop is in those base methods, and the M2 code decided to override those so that the loop could be moved here.

We need to override those methods as well if we're going to loop here.

Reply:
M2 also updates the XML code to use multiple packages, which isn't being done here. So, if we do move to make this class more closely match M2's, then the XML code needs to be changed as well.

Reply:
I don't see how these 2 things are related, XML code will be dead in a few weeks

Reply:
They're related because the behavior of the XML methods should match the behavior of the Rest methods. If one set of methods loops over the request packages, then they both should.

However, I'd argue that neither of them should.

For example, look at the `collectRates()` method. In OpenMage, `collectRates()` is called by ` Mage_Shipping_Model_Shipping::collectCarrierRates`, *inside a foreach* loop.

For the M2 code, the method `Magento\Shipping\Model\Shipping::collectCarrierRates` calls `collectRates()` on the Carrier, but *there is no foreach loop*.

So, for M2, they moved the package loop into the Carrier implementations. It doesn't make sense to loop over the packages in this class unless we're also going to change the other carriers and all of the code that uses carrier models.